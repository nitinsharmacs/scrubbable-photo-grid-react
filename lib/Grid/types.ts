import type { GridConfigType, SectionType } from 'lib/types';

export interface GridProps {
  gridData: SectionType[];
  config: GridConfigType;
}
