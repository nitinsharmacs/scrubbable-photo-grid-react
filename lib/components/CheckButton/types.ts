import type { CSSProperties } from 'react';

export interface CheckButtonProps {
  active?: boolean;
  label: string;
  id: string;
  styles?: CSSProperties;
  onClick?: (e: React.SyntheticEvent) => void;
}
