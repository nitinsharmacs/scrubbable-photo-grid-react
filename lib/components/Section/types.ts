import type { SegmentCtx } from 'lib/components/Segment/types';
import type { GridConfigType, SectionConfigType, SectionType } from 'lib/types';

export type SectionSelectionHandler = (
  sectionId: string,
  ctx: SectionCtx
) => void;

export interface SectionProps {
  section: SectionType;
  config: GridConfigType;
  map: SectionConfigType;
  onSelect?: SectionSelectionHandler;
}

export type SectionCtx = {
  selected: boolean;
  segments: {
    [key: string]: SegmentCtx;
  };
};
