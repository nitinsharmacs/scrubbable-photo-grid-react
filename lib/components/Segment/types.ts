import type { SegmentConfig, SegmentType } from 'lib/types';

export interface SegmentProps {
  segment: SegmentType;
  map: SegmentConfig;
}
