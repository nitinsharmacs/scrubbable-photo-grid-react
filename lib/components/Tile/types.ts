import type { ImageType, TileType } from 'lib/types';

export interface TileProps {
  tile: TileType;
  image: ImageType;
}
