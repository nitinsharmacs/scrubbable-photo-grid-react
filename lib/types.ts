export type SectionType = {
  sectionId: string;
  totalImages: number;
  segments: SegmentType[];
  header: string;
};

export type SegmentType = {
  segmentId: string;
  header: string;
  images: ImageType[];
};

export type TileType = {
  width: number;
  height: number;
  top: number;
  left: number;
  aspectRatio?: number;
};

export type ImageType = {
  metadata: {
    width: number;
    height: number;
    orientation?: number;
  };
  imageurl: string;
  imageId: string;
};

export type SegmentConfigType = {
  top: number;
  width: number;
  height: number;
  tiles: TileType[];
  headerHeight: number;
};

export type SegmentsMapType = {
  [key: string]: SegmentConfigType;
};

export type SectionConfigType = {
  height: number;
  width?: number;
  top: number;
  visible: boolean;
  index: number;
  segmentsMap: SegmentsMapType;
  headerHeight: number;
};

export type SectionsMapType = {
  [key: string]: SectionConfigType;
};

export type GridConfigType = {
  containerWidth: number;
  sectionMargin: number;
  segmentMargin: number;
  targetRowHeight: number;
};
