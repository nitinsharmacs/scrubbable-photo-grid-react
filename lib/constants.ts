export const IMAGE_ASPECT_RATIO: number = 3 / 2;
export const COALESCING_FACTOR: number = 7 / 10;
export const SEGMENT_HEADER_HEIGHT: number = 40;
export const SECTION_HEADER_HEIGHT: number = 50;
