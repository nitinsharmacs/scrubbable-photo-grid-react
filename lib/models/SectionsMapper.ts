import { createSegmentsMap, estimateSectionHeight } from 'lib/helpers';
import { SectionMap } from 'lib/models/SectionMap';
import type {
  GridConfigType,
  SectionsMapType,
  SectionType,
  SectionConfigType,
} from 'lib/types';

export class SectionsMapper {
  private readonly gridConfig: GridConfigType;

  private sectionsMap: SectionsMapType;

  constructor(gridConfig: GridConfigType) {
    this.gridConfig = gridConfig;
    this.sectionsMap = {};
  }

  createMap(sections: SectionType[]): SectionsMapType {
    const map = sections.reduce<SectionsMapType>(
      (map: SectionsMapType, section: SectionType, index: number) => {
        const sectionHeight = estimateSectionHeight(
          section.totalImages,
          this.gridConfig
        );
        return {
          ...map,
          [section.sectionId]: {
            height: sectionHeight,
            top: this.gridConfig.sectionMargin + map['prev'].height,
            visible: false,
            index,
            segmentsMap: {},
          },
          prev: {
            ...map.prev,
            height:
              sectionHeight +
              this.gridConfig.sectionMargin * 2 +
              map['prev'].height,
          },
        } as SectionsMapType;
      },
      {
        prev: { height: 0, top: 0, visible: false, index: -1, segmentsMap: {} },
      }
    );

    delete map['prev'];

    this.sectionsMap = map;

    return { ...this.sectionsMap };
  }

  getMap(): SectionsMapType {
    return { ...this.sectionsMap };
  }

  getSectionConfig(sectionId: string): SectionConfigType {
    return { ...this.sectionsMap[sectionId] };
  }

  updateForSection(
    section: SectionType,
    isVisible: boolean
  ): SectionConfigType {
    const oldSectionMap: SectionConfigType =
      this.sectionsMap[section.sectionId];

    if (!isVisible) {
      oldSectionMap.visible = false;
      return oldSectionMap;
    }

    const newSectionMap: SectionConfigType = this.recomputeSectionMap(section);

    newSectionMap.visible = isVisible;
    this.sectionsMap[section.sectionId] = newSectionMap;

    return newSectionMap;
  }

  private updateSectionsTop(
    toBeUpdatedSectionsId: string[],
    delta: number
  ): SectionsMapType {
    return toBeUpdatedSectionsId.reduce<SectionsMapType>(
      (map, sectionId: string) => {
        return {
          ...map,
          [sectionId]: { ...map[sectionId], top: map[sectionId].top + delta },
        };
      },
      {
        ...this.sectionsMap,
      }
    );
  }

  updateSectionsTopFrom(
    sectionPos: number,
    delta: number,
    sections: SectionType[]
  ): SectionsMapType {
    const nextToBeSectionsId: string[] = sections
      .slice(sectionPos)
      .map((section: SectionType) => section.sectionId);

    this.sectionsMap = {
      ...this.sectionsMap,
      ...this.updateSectionsTop(nextToBeSectionsId, delta),
    };

    return { ...this.sectionsMap };
  }

  recomputeSectionMap(section: SectionType): SectionConfigType {
    const sectionMap: SectionConfigType = this.sectionsMap[section.sectionId];
    const segmentsMap = createSegmentsMap(section.segments, this.gridConfig);

    const sectionMapper = new SectionMap(this.gridConfig, segmentsMap);

    return {
      ...sectionMap,
      segmentsMap: segmentsMap,
      height: sectionMapper.height,
    };
  }
}
