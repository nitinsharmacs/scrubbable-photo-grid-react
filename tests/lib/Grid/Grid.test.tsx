import React from 'react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';

import Grid from 'lib/Grid/Grid';
import { GridConfigType } from 'lib/types';

import sections from '../../data/sections.json';

import { SectionsMapper } from 'lib/models/SectionsMapper';

vi.mock('lib/models/SectionsMapper', () => {
  const SectionsMapperMock = vi.fn();
  SectionsMapperMock.prototype.createMap = vi.fn();
  SectionsMapperMock.prototype.getMap = vi.fn();
  SectionsMapperMock.prototype.getSectionConfig = vi.fn();
  SectionsMapperMock.prototype.updateForSection = vi.fn();
  SectionsMapperMock.prototype.updateSectionsTop = vi.fn();
  SectionsMapperMock.prototype.updateSectionsTopFrom = vi.fn();
  SectionsMapperMock.prototype.recomputeSectionMap = vi.fn();

  return {
    SectionsMapper: SectionsMapperMock,
  };
});

const mockIntersectionObserver = (
  behaviorOverrides = {},
  cb = (_: any) => {}
) => {
  const behavior = {
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
    ...behaviorOverrides,
  };

  const IntersectionObserverMock = vi.fn().mockImplementation((handler) => {
    cb(handler);
    return behavior;
  });
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);

  return behavior;
};

describe('Grid', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  const gridConfig: GridConfigType = {
    containerWidth: 800,
    segmentMargin: 10,
    sectionMargin: 20,
    targetRowHeight: 150,
  };

  let mapper;
  beforeEach(() => {
    mapper = new SectionsMapper(gridConfig);
    vi.mocked(mapper.createMap).mockReturnValue({
      '2019_04': {
        height: 380,
        top: 20,
        visible: true,
        index: 0,
        segmentsMap: {
          '2019_04_25': { top: 10, height: 10, width: 10, tiles: [] },
          '2019_04_24': { top: 10, height: 10, width: 10, tiles: [] },
        },
      },
      '2018_12': {
        height: 150,
        top: 440,
        visible: false,
        index: 1,
        segmentsMap: {},
      },
    });
  });

  it('should match snapshot', () => {
    const observerMock = mockIntersectionObserver();

    const { container } = render(
      <Grid config={gridConfig} gridData={sections} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should observe sections', () => {
    const observeMock = vi.fn();
    const observerMock = mockIntersectionObserver({ observe: observeMock });

    render(<Grid config={gridConfig} gridData={sections} />);

    expect(observeMock).toBeCalledTimes(2);
  });

  it('should update section on intersection', () => {
    const observerMock = mockIntersectionObserver({}, (handler) => {
      handler([{ target: { id: 'someid' }, isIntersecting: true }]);
    });

    vi.mocked(mapper.getSectionConfig).mockReturnValue({
      height: 310,
      top: 20,
      visible: false,
      index: 0,
      segmentsMap: {},
    });
    vi.mocked(mapper.updateForSection).mockReturnValue({
      height: 380,
      top: 20,
      visible: true,
      index: 0,
      segmentsMap: {
        '2019_04_25': { top: 10, height: 10, width: 10, tiles: [] },
        '2019_04_24': { top: 10, height: 10, width: 10, tiles: [] },
      },
    });

    render(<Grid config={gridConfig} gridData={sections} />);

    expect(mapper.getSectionConfig).toBeCalledWith('someid');
    expect(mapper.updateForSection).toBeCalledWith(sections[0], true);
    expect(mapper.updateSectionsTopFrom).toBeCalledWith(1, 70, sections);
  });
});
