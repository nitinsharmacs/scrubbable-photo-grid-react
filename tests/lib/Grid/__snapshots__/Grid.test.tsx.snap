// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grid > should match snapshot 1`] = `
<div>
  <div
    class="grid"
  >
    <div
      class="section"
      id="2019_04"
      style="width: 800px; height: 380px; top: 20px;"
    >
      <div
        class="section-header"
      >
        <h2>
          April, 2019
        </h2>
      </div>
      <div
        class="section-content"
      >
        <div
          class="segment"
          id="2019_04_25"
          style="width: 10px; height: 10px; top: 10px;"
        >
          <div
            class="segment-header"
          >
            <h3>
              Thu, April 25, 2019
            </h3>
          </div>
          <div
            class="segment-content"
          />
        </div>
        <div
          class="segment"
          id="2019_04_24"
          style="width: 10px; height: 10px; top: 10px;"
        >
          <div
            class="segment-header"
          >
            <h3>
              Thu, April 24, 2019
            </h3>
          </div>
          <div
            class="segment-content"
          />
        </div>
      </div>
    </div>
    <div
      class="section"
      id="2018_12"
      style="width: 800px; height: 150px; top: 440px;"
    />
  </div>
</div>
`;
